### DISCUSSION TITLE
Please make TeamTalk compatible with AirPods Pro and Bluetooth speakers by adding a setting to have it not use allowBluetooth but only allowBluetoothA2DP

### DISCUSSION THREAD

**Discussion by XP-Fan** [BearWare/TeamTalk5#1476] [Ideas]
Hey friends,

So first of all a billion thanks, TeamTalk is an extremely amazing software. However, I'd ask you to fix a small annoyance within the iOS version. That is, because it has the audio option allowBluetooth enabled, it would always force the AirPods Pro to go into their crappy low bandwith mono mode and it would also always use the AirPods Pro Microphone rather than keeping the possibility to record in stereo while AirPods Pro are connected. So, could you please give us an option in settings to avoid allowBluetooth and use allowBluetoothA2DP only? Of course the setting would probably need a less iOS freak like name, guess something like "Avoid mono bluetooth" would work. With this setting enabled it would then no longer use mono bluetooth gear or bluetooth microphones, but in return it could use AirPods Pro in stereo while recording from the iPhone's internal microphones and it could also stream to bluetooth speakers. I hope I managed to explain this clearly.

Thanks in advance!

**Comment by FranceBB** [comment]
Hi there,
one of my friends actually has the same problem.
I mean, it's not a problem per se, but she let me use her AirPods Pro with her iPhone 13 Pro once when I forgot my Google Pixel 6 Pro at home and while the quality of media etc was ok, when she tried to stream back via TeamTalk I could hear it falling back to some other standard (which I thought was SBC at first, but then I realized it was some proprietary mono shit by Apple). Although it might work for things like meetings etc, it was awful for pretty much everything else, especially music when you compare it to the Advanced Audio Distribution Profile.
Anyway, it would be nice if you could take a look into this and offer an option to users to choose one or the other on supported devices.

Thank you for making this software free, open source and cross-platform (I'm a developer of other open source projects myself and I can always say: open source rocks. :) )

Cheers,
Frank

**Comment by Simon818** [comment]
I've wanted to suggest something similar for some time. I think if a microphone other than the Bluetooth headset is selected, the headset should switch to A2DP mode. It's immensely frustrating having an expensive phone with great microphones, expensive AirPods with great sound, and not being able to use either to their full potential. I don't know how complicated this is to implement, but it would be as much of a game changer as the simultaneous audio playback trick—in fact, the two would work together and allow us to listen to full quality music and other audio while still connected to TeamTalk. TeamTalk is easily the biggest reason I find myself using wired headphones these days, as the lag of AirPods is minimal even in stereo mode.

**Comment by XP-Fan** [comment]
Exactly and the worst part is: Nothing changes on that situation with the new AirPods Pro 2th gen. I was hoping that they would use the LC3 codec to allow for at least basic quality stereo receiving and a slight quality increase on the microphone side, but no, we still get and send crappy mono, 24 KHZ AAC-ELD, with our sending part even being run through an extremely low quality noise cancellation that ensures that only human voice is transmitted. While I know that this is not TeamTalks fault (you didn’t design the new AirPods), it does however make it even more important that we have an option to avoid those crappy microphones altogether, as literally everyone, with any generation of AirPods, is affected, so please please please make this happen!

Von: Simon Jaeger ***@***.***
Gesendet: Dienstag, 27. September 2022 11:58
An: BearWare/TeamTalk5 ***@***.***>
Cc: XP-Fan ***@***.***>; Author ***@***.***>
Betreff: Re: [BearWare/TeamTalk5] Please make TeamTalk compatible with AirPods Pro and Bluetooth speakers by adding a setting to have it not use allowBluetooth but only allowBluetoothA2DP (Discussion #1476)


I've wanted to suggest something similar for some time. I think if a microphone other than the Bluetooth headset is selected, the headset should switch to A2DP mode. It's immensely frustrating having an expensive phone with great microphones, expensive AirPods with great sound, and not being able to use either to their full potential. I don't know how complicated this is to implement, but it would be as much of a game changer as the simultaneous audio playback trick—in fact, the two would work together and allow us to listen to full quality music and other audio while still connected to TeamTalk. TeamTalk is easily the biggest reason I find myself using wired headphones these days, as the lag of AirPods is minimal even in stereo mode.

**Comment by bear101** [comment]
@XP-Fan If I remove allowBluetooth and only pass "allowAirPlay, allowBluetoothA2DP" then I don't see the Air Pods as an available sound input. I only see the iPhone's microphones (Bottom, Front, Back). Is this acceptable behaviour?

Also from the [A2DP documentation](https://developer.apple.com/documentation/avfaudio/avaudiosession/categoryoptions/1771735-allowbluetootha2dp) it states: "A2DP is a stereo, output-only profile intended for higher bandwidth audio use cases, such as music playback".

**Reply by Simon818** [reply]
Would this mean that the AirPods wouldn't be available as an input under any circumstances? I think there were beta builds I didn't test, so sorry for jumping in with slightly incomplete information.

**Comment by XP-Fan** [comment]
Sorry for my late response I was on a wedding. Well, that’s correct: Technically you can’t really use the AirPods as input without automatically forcing them to fall into crappy 24 KHZ mono AAC-ELD with both input and output, which I personally perceive as simply unusable. For Simon, yes, it would mean that the AirPods can no longer be used as inputs, but again for me that would be totally fine. If it’s not for you then the best suggestion I can currently come up with would be another checkbox within TeamTalk settings maybe?

**Comment by Simon818** [comment]
Well, there is a significant use case for talking on AirPods; they sound no worse than they do on a FaceTime call and people use them for that, including myself. I think if that functionality were to be removed people would care a lot, as it would mean they could no longer use TeamTalk hands-free. Good quality is great and all, but a lot of people just use TeamTalk for--well--talking, and sometimes the convenience of being hands-free is more important than having high-quality conversations.

However, it seems like the solution on the user interface side is simple. (I'm not sure how easy the coding side is.) Right now, while AirPods are in use, selecting a microphone other than AirPods will force the output to go back to the phone as well. Instead of doing this, why not just transparently switch to A2DP whenever the active Bluetooth device's microphone is not in use? This is far more intuitive and doesn't require any configuring our thought on the user's part. If they want to use the AirPods mic, they'll sacrifice some quality but it will work. If they want to use some other mic, they can take full advantage of the nice stereo A2DP audio that is now possible. No additional configuration option is necessary because there is never a reason to use the phone profile instead of A2DP on a modern Bluetooth device, unless the microphone of said device is also in use.

**Comment by XP-Fan** [comment]
Hm, okay. I would never torture anyone with 24 KHZ, though. However your suggestion sounds sane.

**Comment by Simon818** [comment]
I noticed the new beta with the A2DP change; thank you for that. I'm testing and so far it seems to work really well. With A2DP toggled on, I have stereo audio from TeamTalk, stereo audio from the rest of the phone, and can even play audio in the background while on a server. This is a pretty huge deal and something I don't think any other voice conferencing app allows over Bluetooth.

This seems to have broken stereo microphones: If I select "Back" or "Front" as the preferred microphone (with or without Bluetooth), it only transmits the top front or back microphone in both channels.

**Reply by ChrisDuffley** [reply]
@Simon818, yes, I've seen this happen a lot since at least the beginning of iOS 16, but I cannot say when exactly this started. If I'm not mistaken, I think other apps are doing this as well. RecorderHQ, for instance, also has this problem and you have to restart the phone just to get stereo working again in 3rd-party apps. Hope this helps clarify!
